### DISCUSSION TITLE
EarthManipulator setViewPoint Thread Safety

### DISCUSSION THREAD

**Discussion by nebsar** [pelicanmapping/osgearth#2009] [Support Q&A] [answered]
I am trying to update a vehicle position on the fly, based on a recorded log data (some kind of playback). I center view on the vehicle and tether the view by setting EarthManipulitor::setViewPoint. I am using OpenThreads for this, but I am not familiar with OpenThreads. After some time the app naturally crashes. But if I do not center the view and do not follow the vehicle, app does not crash. I am not familiar with the thread safety on OpenSceneGraph and osgEarth. What should I do to achieve thread safety on EarthManipulitor::setViewPoint? The code I use in the thread is below:

virtual void run() override {
		while (!_stop) {
			YieldCurrentThread();
			DataLine dataLine = _vData.at(_index++);
			osgEarth::GeoPoint vPosition(_srs, dataLine.longitude, dataLine.latitude, dataLine.altitude, osgEarth::AltitudeMode::ALTMODE_ABSOLUTE);
			osgEarth::Viewpoint vp = _manipulator->getViewpoint();
			osg::DegreesToRadians(dataLine.roll), osg::Vec3(0, 1, 0), osg::DegreesToRadians(dataLine.trueHeading), osg::Vec3(0, 0, -1)));
			vp.setNode(_trackNode->getGeoTransform());
			vp.setHeading(osgEarth::Angle(dataLine.heading, osgEarth::Units::DEGREES));
			_manipulator->setViewpoint(vp);
			_trackNode->setPosition(vPosition);
			_index++;
			microSleep(40000);
		}
	}

**Comment by jasonbeverage** [comment, marked as answer]
You need to set the viewpoint in the update phase of the scene graph loop
instead of in your thread.

**Reply by nebsar** [reply]
Thank you Jason. Where is the update phase of the scene graph loop? I am really new to updates on osg and osgEarth?

**Reply by nebsar** [reply]
Is it some kind of a camera callback?

**Reply by nebsar** [reply]
Ok I started to update the setViewPoint in camera callback. I think it did the trick! Working without a crash now. I learned a new thing. Thanks man.

**Comment by jasonbeverage** [comment]
Perfect!  Glad it worked.  In general an osg frame loop has 4 phases.
event, update, cull and draw.  It's generally only safe to modify the scene
graph (or do most things unless you really know what you are doing) in the
event or update phase.  The event phase handles mouse movement, key
presses, joysticks, etc and is usually handled by osgGA::EventAdapters.
The update phase runs any UpdateCallbacks you have on your nodes or Update
Operations.

Jason
